core/src/impl/Kokkos_Stacktrace.cpp

```diff
@@ -0,0 +1,265 @@
+#include "Kokkos_Stacktrace.hpp"
+
+// backtrace() function for retrieving the stacktrace
+#include <execinfo.h>
+
+#define KOKKOS_ENABLE_CXXABI 1
+
+#ifdef KOKKOS_ENABLE_CXXABI
+#  include <cxxabi.h>
+#endif // KOKKOS_ENABLE_CXXABI
+
+#include <exception>
+#include <iostream>
+#include <tuple>
+#include <vector>
+
+namespace Kokkos {
+namespace Impl {
+
+std::string demangle (const std::string& name)
+{
+#ifndef KOKKOS_ENABLE_CXXABI
+  return name;
+#else
+  std::string s;
+
+  if (name.length () != 0) {
+    int status = 0;
+    char* output_buffer = nullptr;
+    size_t* length = nullptr;
+    char* d = abi::__cxa_demangle (name.c_str (), output_buffer, length, &status);
+    if (d != nullptr) {
+      s = d;
+      free (d);
+    }
+  }
+
+  // Special cases for "main" and "start" on Mac
+  if (s.length () == 0) {
+    if (name == "main" || name == "start") {
+      s = name;
+    }
+  }
+#endif // KOKKOS_ENABLE_CXXABI
+
+  return s;
+}
+
+class Stacktrace {
+public:
+  Stacktrace () = delete;
+  Stacktrace (const Stacktrace&) = delete;
+  Stacktrace& operator= (const Stacktrace&) = delete;
+  Stacktrace (Stacktrace&&) = delete;
+  Stacktrace& operator= (Stacktrace&&) = delete;
+  ~Stacktrace () = delete;
+
+  // These are public only to avoid wasting an extra stacktrace line.
+  // See save_stacktrace below.
+  static constexpr int capacity = 100;
+  static void* buffer[capacity];
+  static int length;
+
+  static std::vector<std::string> lines ()
+  {
+    char** symbols = backtrace_symbols (buffer, length);
+    if (symbols == nullptr) {
+      return {};
```

Comment:
Is that the right thing to do on error?  Should we throw instead?

Reply:
Given that the stacktrace code is executed from signal handlers or exception handlers, I don't think it should throw exceptions.

Reply:
I agree we shouldn't throw.  `backtrace_symbols` is not safe to call from a signal handler, though.